scripts/question_answering/utils.py

```diff
@@ -0,0 +1,379 @@
+        logconsole.setFormatter(formatter)
+        logging.root.addHandler(logconsole)
+
+    return folder
+
+
+def get_very_negative_number():
+    return -1e30
+
+
+def extend_to_batch_size(batch_size, prototype, fill_value=0):
+    """Provides NDArray, which consist of prototype NDArray and NDArray filled with fill_value to
+    batch_size number of items. New NDArray appended to batch dimension (dim=0).
+
+    Parameters
+    ----------
+    batch_size: ``int``
+        Expected value for batch_size dimension (dim=0).
+    prototype: ``NDArray``
+        NDArray to be extended of shape (batch_size, ...)
+    fill_value: ``float``
+        Value to use for filling
+    """
+    if batch_size == prototype.shape[0]:
+        return prototype
+
+    new_shape = (batch_size - prototype.shape[0],) + prototype.shape[1:]
+    dummy_elements = nd.full(val=fill_value, shape=new_shape, dtype=prototype.dtype,
+                             ctx=prototype.context)
+    return nd.concat(prototype, dummy_elements, dim=0)
+
+
+def get_combined_dim(combination, tensor_dims):
+    """
+    For use with :func:`combine_tensors`.  This function computes the resultant dimension when
+    calling ``combine_tensors(combination, tensors)``, when the tensor dimension is known.  This is
+    necessary for knowing the sizes of weight matrices when building models that use
+    ``combine_tensors``.
+
+    Parameters
+    ----------
+    combination : ``str``
+        A comma-separated list of combination pieces, like ``"1,2,1*2"``, specified identically to
+        ``combination`` in :func:`combine_tensors`.
+    tensor_dims : ``List[int]``
+        A list of tensor dimensions, where each dimension is from the `last axis` of the tensors
+        that will be input to :func:`combine_tensors`.
+    """
+    combination = combination.replace('x', '1').replace('y', '2')
+    return sum([_get_combination_dim(piece, tensor_dims) for piece in combination.split(',')])
+
+
+def _get_combination_dim(combination, tensor_dims):
+    if combination.isdigit():
+        index = int(combination) - 1
+        return tensor_dims[index]
+    else:
+        first_tensor_dim = _get_combination_dim(combination[0], tensor_dims)
+        return first_tensor_dim
+
+
+def _get_combination(combination, tensors):
+    if combination.isdigit():
+        index = int(combination) - 1
+        return tensors[index]
+    else:
+        first_tensor = _get_combination(combination[0], tensors)
+        second_tensor = _get_combination(combination[2], tensors)
+        operation = combination[1]
+        if operation == '*':
+            return first_tensor * second_tensor
+        elif operation == '/':
+            return first_tensor / second_tensor
+        elif operation == '+':
+            return first_tensor + second_tensor
+        elif operation == '-':
+            return first_tensor - second_tensor
+        else:
+            raise NotImplementedError
+
+
```

Comment:
Would it make sense to make the functions below `HybridBlocks`?

Reply:
Don't know, will talk with Chenguang about that. If yes, then will do it before merging to master.